src/rust/src/environment.rs

```diff
@@ -3,31 +3,60 @@ use arrow::array::RecordBatchReader;
 use arrow::ffi_stream::{ArrowArrayStreamReader, FFI_ArrowArrayStream};
 use datafusion::arrow::array::RecordBatch;
 use datafusion::datasource::{MemTable, TableProvider};
+use savvy::ffi::SEXP;
+use savvy::protect::{insert_to_preserved_list, release_from_preserved_list};
+use savvy::{EnvironmentSexp, Sexp};
 use sqlexec::environment::EnvironmentReader;
-use std::sync::Arc;
+use std::sync::{Arc, Mutex};
 
-#[derive(Debug, Clone, Copy)]
-pub struct REnvironmentReader;
+// TODO
+struct UnsafeToken(SEXP);
+unsafe impl std::marker::Send for UnsafeToken {}
+unsafe impl std::marker::Sync for UnsafeToken {}
+struct UnsafeEnvironmentSexp(EnvironmentSexp);
+unsafe impl std::marker::Send for UnsafeEnvironmentSexp {}
+unsafe impl std::marker::Sync for UnsafeEnvironmentSexp {}
+
+pub struct REnvironmentReader {
+    env: Arc<Mutex<UnsafeEnvironmentSexp>>,
+    token: UnsafeToken,
+}
+
+impl REnvironmentReader {
+    pub(crate) fn new(env: EnvironmentSexp) -> Self {
+        let token = insert_to_preserved_list(env.inner());
+        Self {
+            env: Arc::new(Mutex::new(UnsafeEnvironmentSexp(env))),
+            token: UnsafeToken(token),
+        }
+    }
+}
+
+impl Drop for REnvironmentReader {
+    fn drop(&mut self) {
+        release_from_preserved_list(self.token.0)
+    }
+}
```

Comment:
@yutannihilation Sorry to bother you.
This package is AGPL-3.0 licensed, but may I copy this to https://github.com/pola-rs/r-polars and distribute it under the terms of the MIT license (re-licensing)?

Reply:
Yes, please feel free to use with any license! (I'm not fully confident if this is the best option, though)

Reply:
Thanks!
Thanks for your help, R udf worked.
https://github.com/eitsupi/neo-r-polars/commit/a9893624feb6087bca8412cedc8558a67c1af48a

Reply:
Great progress!